components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,402 @@
+    return false;
+  }
+  return true;
+}
+
+bool FillStringField(StringTokenizer* tokenizer,
+                     const std::string& input,
+                     const std::string& msg_token,
+                     bool consume_delim,
+                     std::string& field) {
+  CHECK(tokenizer);
+  if (!CheckField(input, msg_token)) {
+    return false;
+  }
+  field = input.substr(msg_token.length());
+  if (field.empty()) {
+    return false;
+  }
+  if (consume_delim && !ConsumeDelim(tokenizer)) {
+    return false;
+  }
+  return true;
+}
+
+bool FillOptionalStringField(StringTokenizer* tokenizer,
+                             const std::string& msg_token,
+                             absl::optional<std::string>& field) {
+  CHECK(tokenizer);
+  std::string field_str;
+  if (!FillStringField(tokenizer, tokenizer->token(), msg_token, false,
+                       field_str)) {
+    return false;
+  }
+  field = field_str;
+  return true;
+}
+
+bool FillOptionalResourcesField(StringTokenizer* tokenizer,
+                                absl::optional<std::vector<GURL>>& field) {
+  CHECK(tokenizer);
+  if (!CheckField(tokenizer->token(), kResourcesToken)) {
+    return false;
+  }
+  std::vector<GURL> urls;
+  while (tokenizer->GetNext()) {
+    if (tokenizer->token_is_delim()) {
+      continue;
+    }
+    if (!CheckField(tokenizer->token(), kResourcesSeperator)) {
+      return false;
+    }
+    auto url_str = tokenizer->token().substr(strlen(kResourcesSeperator));
+    GURL url(url_str);
+    if (!url.is_valid()) {
+      return false;
+    }
+    urls.push_back(url);
+  }
+  if (!urls.size()) {
+    return false;
+  }
+  field = urls;
+  return true;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-2.3
+bool IsUnreservedChar(char c) {
+  if (base::IsAsciiAlphaNumeric(c)) {
+    return true;
+  }
+  switch (c) {
+    case '-':
+    case '.':
+    case '_':
+    case '~':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986#section-2.2
+bool IsSubDelimChar(char c) {
+  switch (c) {
+    case '!':
+    case '$':
+    case '&':
+    case '\'':
+    case '(':
+    case ')':
+    case '*':
+    case '+':
+    case ',':
+    case ';':
+    case '=':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-3.3
+bool IsPChar(char c) {
+  if (IsUnreservedChar(c)) {
+    return true;
+  }
+  if (IsSubDelimChar(c)) {
+    return true;
+  }
+  switch (c) {
+    case ':':
+    case '[':
+    case ']':
+    case '@':
+      return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+mojom::SIWEMessagePtr SIWEMessageParser::Parse(const std::string& message) {
+  state_ = State::kStart;
+  StringTokenizer tokenizer(message, "\n");
+  tokenizer.set_options(StringTokenizer::RETURN_DELIMS);
+  auto result = mojom::SIWEMessage::New();
+  bool seeking_optional_fields = false;
+  while (tokenizer.GetNext() || seeking_optional_fields) {
```

Comment:
I'm seeing some lagginess within the UI and not sure if it's because of the parsing, UI, or just the development build. I'm thinking this is an issue of just running this on a development build, but is it possible that this while loop isn't operating in time complexity of O(n) where n is the number of tokens to be parsed?

Mainly just asking because I've not seen this code pattern (switch in while loop) used very often and wasn't sure if the compiler might be doing something weird in C++ with it.

Reply:
That is just development build, looping through the result split by `\n` is O(n)